packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -18,13 +18,176 @@ import 'build_info.dart';
+      vmServices[i] = VMService.connect(observatoryUris[i]);
+      printTrace('Connected to service protocol: ${observatoryUris[i]}');
+    }
+  }
+
+  Future<Null> refreshViews() async {
+    if ((vmServices == null) || vmServices.isEmpty)
+      return;
+    for (VMService service in vmServices)
+      await service.vm.refreshViews();
+    _viewsCache = null;
+  }
+
+  List<FlutterView> _viewsCache;
+  List<FlutterView> get views {
+    if (_viewsCache == null) {
+      if ((vmServices == null) || vmServices.isEmpty)
+        return null;
+      final List<FlutterView> result = <FlutterView>[];
+      if (_viewFilter == null) {
+        for (VMService service in vmServices) {
+          if (!service.isClosed)
+            result.addAll(service.vm.views.toList());
+        }
+      } else {
+        for (VMService service in vmServices) {
+          if (!service.isClosed)
+            result.addAll(service.vm.allViewsWithName(_viewFilter));
+        }
+      }
+      _viewsCache = result;
+    }
+    return _viewsCache;
+  }
+
+  Future<Null> getVMs() async {
+    for (VMService service in vmServices)
+      await service.getVM();
+  }
+
+  Future<Null> waitForViews() async {
+    // Refresh the view list, and wait a bit for the list to populate.
+    for (VMService service in vmServices)
+      await service.waitForViews();
+  }
+
+  Future<Null> stopApps() async {
+    final List<FlutterView> flutterViews = views;
+    if (flutterViews == null || flutterViews.isEmpty)
+      return;
+    for (FlutterView view in flutterViews) {
+      if (view != null && view.uiIsolate != null)
+        view.uiIsolate.flutterExit();
+    }
+    await new Future<Null>.delayed(const Duration(milliseconds: 100));
+  }
+
+  Future<Uri> setupDevFS(String fsName,
+                         Directory rootDirectory, {
+                         String packagesFilePath
+                       }) {
+    // One devFS per device. Shared by all running instances.
+    devFS = new DevFS(
+      vmServices[0],
+      fsName,
+      rootDirectory,
+      packagesFilePath: packagesFilePath
+    );
+    return devFS.create();
+  }
+
+  List<Future<Map<String, dynamic>>> reloadSources(String entryPath, {
+                                                   bool pause: false
+                                                 }) {
+    final Uri deviceEntryUri = devFS.baseUri.resolveUri(fs.path.toUri(entryPath));
+    final Uri devicePackagesUri = devFS.baseUri.resolve('.packages');
+    final List<Future<Map<String, dynamic>>> reports = <Future<Map<String, dynamic>>>[];
+    for (FlutterView view in views) {
+      final Future<Map<String, dynamic>> report = view.uiIsolate.reloadSources(
+        pause: pause,
+        rootLibUri: deviceEntryUri,
+        packagesUri: devicePackagesUri
+      );
+      reports.add(report);
+    }
+    return reports;
+  }
+
+  Future<Null> debugDumpApp() async {
+    for (FlutterView view in views)
+      await view.uiIsolate.flutterDebugDumpApp();
```

Comment:
This (and the next one) seem unlikely to be useful to the user, but I don't have a better suggestion.
(Dumping one tree is bad enough, dumping multiple trees... yikes. I guess at least they won't overlap.)